rfcs/20230606-topk.md

```diff
@@ -0,0 +1,88 @@
+# RFC: Add TopK Op to StableHLO
+
+Status: Draft<br/>
+Initial version: 2023-06-06<br/>
+Last updated: 2023-06-06
+
+## Summary
+
+We propose to add a new op: `stablehlo.topk`, which returns the first `k`
+elements of a tensor and their indices, based on a user-provided comparator. The
+goal of the new op is to bridge a gap between existing and emerging uses of TopK
+in models (especially transformers and LLMs, e.g.: [^1], [^2]), as well as in
+the IR lowering chain.
+
+## Background and Motivation
+
+The TopK operation is ubiquitous in both ML frameworks (e.g.,
+[JAX](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.top_k.html),
+[tensorflow](https://www.tensorflow.org/api_docs/python/tf/math/top_k),
+[PyTorch](https://pytorch.org/docs/stable/generated/torch.topk.html)) and
+compiler IRs ([XLA](https://www.tensorflow.org/xla/operation_semantics#top-k),
+[CHLO](https://github.com/openxla/stablehlo/blob/6f0d25ef2e05f4549c5e4ffc506436a4a13af51f/stablehlo/dialect/ChloOps.td#L811-L842),
+[MHLO](https://github.com/openxla/xla/commit/11a4424760f8814faa6da29a0e6dd9d5c168d22a#diff-9573e4e3a2bf686edee3437726f9dbccaf1b296e634d44f4df504d94bf2576d1),
+[LinalgExt](https://github.com/openxla/iree/blob/b5e45cd08eecbd97b0568f9c5fdac28b0f5e8f11/llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td#L436-L480)),
+which makes its absence from StableHLO a point of friction. For example, in IREE
+we have to pattern match stablehlo `iota` + `sort` + `slice` and raise it back
+to `chlo.top_k`, so that we can efficiently lower it to `linalg_ext.topk`. See
+the IREE issue for more details: <https://github.com/openxla/iree/issues/13905>.
+
+The definitions of TopK in frontend frameworks and IRs vary in two places:
+
+* Whether `k` is static or dynamic.
+* Whether the results are sorted, unordered, or whether ordering is decided based
+  on a function argument.
+
+Note that when the input TopK gets lowered to `sort` followed by a dynamic slice,
+we generally lose the information about both. If `k` is a runtime value, we no
+longer know whether we can expect it to be much smaller than the number of
+elements, which makes it difficult to tell whether to rewrite sort as TopK or
+whether we truly have to sort all elements. Second, we may be introducing
+unnecessary ordering when the input semantics do not require for the results to
+be sorted.
+
+Note that we want to avoid degrading to a general `sort`, as this can lead to
+significant bottlenecks in real-world models, e.g.:
+<https://github.com/openxla/iree/issues/13865>.
+
+This loss of high level semantic information and high friction along the
+lowering chain makes us argue for a dedicated TopK op in StableHLO.
```

Comment:
In the recent discussions around MHA, we were floating the idea of having a different mechanism for ops that can be decomposed (like special function call) to preserve them as subgraph but providing any StableHLO compiler a straightforward way to match it to a first class op internally.
Wouldn't that apply here as well?

Reply:
Do you have some link with a reference to these discussions?

Reply:
This is a pain point StableHLO representation should solve.  And I don't think this should lead to `pollution` of arbitrary ops into the StableHLO op set.  Let me give you an example that I recently ran into:

Gaussian Error Linear Unit (GELU) https://arxiv.org/pdf/1606.08415.pdf is an activation layer we needed to support.  It's used in some other text models as well, so it's not a special case.  In the current status, we would have used the first approximation listed in the paper which would get decomposed into 7 ops.  It was performance critical, yes, it was important to us, yes, but do I think it should be added to the StableHLO op set?  My answer would be a no.  The user should do the graph matching and do a custom op for this case if they really care about performance.  Without this, there will be a GELU (original), GELU_V2 (first approximation), and GELU_V3 (second approximation), because one version didn't work for everybody (there are some numerical differences, making the network produce worse output).

GELU was one example.   If we start allowing one performance op into the builtin op set, we can't say no to any other RFCs introducing custom ops into the StableHLO op set.  Someone will come with a hash table look up op and another would want a k-means op.

What StableHLO must resolve is the preserving the high level semantic information mentioned here, so that it's easy for the users to write custom acceleration.

Reply:
Some links:
https://github.com/pytorch/pytorch/blob/main/torch/_decomp/decompositions.py
https://github.com/llvm/torch-mlir/blob/2480cb7a512d249a0cc41b2079313b5ada3c399d/python/torch_mlir/dynamo.py#L39
https://github.com/llvm/torch-mlir/blob/2480cb7a512d249a0cc41b2079313b5ada3c399d/include/torch-mlir/Dialect/Torch/Transforms/Passes.h#L45

Reply:
> a winning strategy here is to declare part of StableHLO the "primitive"/"core" subset and have "decompositions" that can turn any non-primitive op into a combination of core ops.

Fully agree with this! @burmako and I iterated on a design that allows dialects to define StableHLO decompositions, and would appreciate any feedback: [[RFC] StableHLO Extensibility](https://docs.google.com/document/d/1bSyyLA-p1F7KjZgjo563F1WFsPwcZc4eaH5WyQfbsi0/edit#heading=h.kfv34azf3j5k)

This doc proposes two StableHLO interfaces that arbitrary dialects / ops can implement that would allow for backends to specify supported non-primitive ops _at the time of deserialization_. If a non-primitive op is unsupported by a backend, a StableHLO decomposition is used instead.

> We would then have a standardized library of such decompositions that users of StableHLO can select.

This is a separate design point worth discussing, if there should be one canonical extension dialect, or several extension dialects. The proposed design is for arbitrary dialects, but with slight modification would work for a single dialect.

Reply:
I would try to create an ecosystem around this. What I'm seeing from Torch-MLIR is that a lot of people see a lot of value in StableHLO outside the stable artifact itself. So in addition to the core value of the stable artifact, an orthogonal direction is serving as a "lingua franca" for ML compilation, which often will not have all the same constraints as the stable artifact.

Reading Kevin's proposal, it seems like a great idea with the serialized decomposition inline in the module which allows implementations that are unaware of the custom op at least correctly execute :+1:

One design point I'm not sure about is the idea that the proxy is a proxy for a specific op in another dialect, or whether the "custom operation" is abstractly defined independent of any other dialect. I'm not sure that "MLIR dialects" are a strong enough backplane for defining an ecosystem. Probably some org.mycompany.Foo type scheme would be better. Or a Khronos style central registry with vendor proposals that get mixed and matched by vendors and eventually rolled into the next version of the standard if successful. That seems like it would be more durable and build more consensus and glide path to adoption.

I think the balance here is defining
1) a way for vendors to tunnel through arbitrary operations in a well-defined way that is lightweight (is asking for an HLO reference too much to ask for? What about stuff that is not expressible as HLO, such as some linked list thing or other weirdness). Idea: java-style org.mycompany.Foo improved, more-disciplined version of custom_call
2) a way for the community to gradually coalesce around more stable and winning designs. Idea: Khronos style extensions
3) The "official StableHLO"
4) The primitive subset of "official StableHLO"

Ideally we provide a glide path for each operation to graduate core-wise to the maximal level that makes sense.

Note that one of the core value-adds of StableHLO is that it is a small set so that it is plausible to write "a X rule for each operator" (e.g. partitioning), but we have seen that e.g. for TopK it is actually valuable to partition the non-decomposed version, which is more work (have to write a new rule), but better performance, so worth it.

The name of the game here is gradients, allowing things to be stood up and more work invested for greater interop and performance as needed.

Reply:
I had similar questions and doubts to @silvasean but decided to continue on openxa-discuss, at this is not directly related to the TopK RFC itself: https://groups.google.com/a/openxla.org/g/openxla-discuss/c/Ao5K8fvXoEk/m/JG0DGzhqAgAJ

Reply:
This sounds very promising. The initial design I proposed for composites last September was basically "custom_call with default implementation", and the primary feedback was that this may cause ecosystem fragmentation and we should look for an approach that leads to convergence. I think backing this with a public spec may be one of the missing pieces: It would allow backends to safely optimize, community members to reuse existing specs, and frameworks don't need additional dependencies as long as StableHLO provides a means of implementing the spec (maybe functions+metadata, maybe custom_call++).

One possible option: Allow any non-stable abstractions to be made for experimentation, but only permit specced features to appear in portable artifacts.